### PR TITLE
perf(finetune): make Unsloth builds reproducible and cacheable

### PR DESCRIPTION
**Pull request by sozercan**
Builds on the Unsloth compatibility update from #814, which is already merged.

## Summary

- Pin the Python fine-tuning environment to Unsloth/Unsloth Zoo `2026.8.1`, Torch `2.10.0` on CUDA 12.6, and uv `0.12.1`.
- Install the 109-package, hash-locked `pylock.toml` into an isolated virtual environment using a repository-pinned resolution cutoff.
- Replace manual NVIDIA driver installation, device-node creation, and `security.insecure` execution with BuildKit CDI device requests.
- Split Unsloth training and GGUF export into separate BuildKit operations with intentional cache boundaries.
- Resolve and persist an immutable Hugging Face revision for the non-quantized export base, then export from that exact local snapshot.
- Normalize omitted or null fine-tune sections while preserving explicitly configured zero and false values.
- Validate model, dataset, training, optimizer, scheduler, quantization, output-name, NVIDIA driver cache-key, and CDI-device inputs before expensive GPU work.
- Refactor the Unsloth entrypoint into independently testable train and export phases.
- Add smoke and full GPU workflow modes with isolated resources, CDI preflight checks, deterministic configuration rendering, targeted cleanup, inference verification, and diagnostic artifacts.

## Why

The previous fine-tuning path installed a mutable Python dependency set into the system environment, coupled training and GGUF export cache invalidation, manually installed NVIDIA driver userspace components, created GPU device nodes with `mknod`, and required unrestricted BuildKit execution.

This change makes the Python dependency closure reproducible, grants GPU access through narrowly scoped CDI devices, and avoids invalidating expensive training work when only the output name or quantization changes.

## CDI GPU execution

Training and export request an NVIDIA CDI device through `llb.AddCDIDevice` and run with normal BuildKit sandbox security. AIKit does not install an NVIDIA driver in the build container, create `/dev/nvidia*` nodes, or require `--allow security.insecure`.

AIKit defaults to `nvidia.com/gpu=0`. A different NVIDIA CDI selector can be supplied through the `cdiDevice` build argument. The workflow prefers a canonical builder-advertised `GPU-...` or `MIG-...` UUID, while retaining valid on-demand, index, `all`, and type-index fallbacks. Docker runtime inference uses a separate concrete selector, so a bare BuildKit on-demand kind is never passed to `docker run`.

Cross-session GPU result caching is enabled only when an immutable canonical GPU/MIG UUID is paired with the NVIDIA driver version. Mutable selectors fall back to the BuildKit session ID, preventing cache reuse across different hardware while retaining persistent package, model, dataset, and compiler caches.

## Cache behavior

Persistent caches are retained for:

- apt packages and package indexes
- uv downloads
- Hugging Face models
- Hugging Face datasets
- Torch
- Triton
- llama.cpp and Unsloth export tooling

Training and GGUF export are separate graph vertices:

- Changing only `output.name` reuses training and export.
- Changing only `output.quantize` reuses training and reruns export.
- Changing model, dataset, or training configuration reruns training and export.
- Changing the immutable GPU identity or driver invalidates GPU result layers without invalidating OS or Python dependency installation.

Training records the immutable export-base repository revision in the saved PEFT configuration. Export validates that revision, downloads the exact snapshot, rewrites the adapter to reference the local snapshot, and reloads with `local_files_only=True`.

## Validation

Current PR head: `26b2f53e9addb83d8c79d7ffd05ab0592c3865df`

- Required PR checks: **44 passed, 1 expected Dependabot check skipped, 0 failed, 0 pending**.
- `make test` passed with the race detector.
- golangci-lint `v2.11.2`: **0 issues**.
- Python fine-tune suite: **20 tests passed**.
- Unsloth lock verification resolved the same **109 packages** and passed byte-for-byte comparison.
- actionlint/ShellCheck and `git diff --check` passed.
- Full GPU workflow passed end to end in **5m21s**: https://github.com/kaito-project/aikit/actions/runs/30862298790
- Smoke GPU workflow passed end to end in **5m58s**: https://github.com/kaito-project/aikit/actions/runs/30862647614

Both GPU workflows covered CDI preflight, fine-tuning, immutable export-base resolution, GGUF export, artifact verification, custom image build, CDI runtime inference, cleanup, and artifact publication.

## Remaining provenance work

The Python dependency closure and the export base used within a training/export build are pinned. Fresh builds can still observe changes when users supply unversioned remote model or dataset identifiers, and apt packages are resolved from the configured Ubuntu repositories.

Complete clean-build reproducibility would require first-class model and dataset revision fields and, if required, a dated or otherwise immutable OS package source.

### CODE REVIEW

File: scripts/update-unsloth-lock.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+compile_lock() {
+  local directory=$1
+
+  (
+    cd "$directory"
+    uv pip compile requirements.in \
+      --torch-backend=cu126 \
+      --python-version 3.10 \
+      --python-platform x86_64-manylinux_2_28 \
+      --only-binary=:all: \
```
[the remark this answers is not in the document]
Comment (sozercan) [reply scripts/update-unsloth-lock.sh:32]: Fixed in `6f4e922`. The lock generator now uses the repository-pinned cutoff `2026-08-03T00:00:00Z` via `--exclude-newer`, the lock was regenerated, and CONTRIBUTING documents that advancing the cutoff must be deliberate. `make check-unsloth-lock` resolves the same 109 packages and passes.

File: pkg/finetune/target_unsloth.py [hunk 2]
```diff
@@ -52,56 +27,278 @@
...
+    config_text: str,
+    *,
+    loader: Callable[[str], Mapping[str, Any]] | None = None,
+) -> Mapping[str, Any]:
+    if loader is None:
+        import yaml
+
+        loader = yaml.safe_load
+
+    return loader(config_text)
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Fixed in `6f4e922`. `parse_config()` now validates that the loader result is a `Mapping` and raises a clear `ValueError` containing only the invalid root type. Added coverage for empty/null, sequence, and scalar roots; the Python suite now passes 18 tests.

File: pkg/build/build.go [hunk 3]
```diff
@@ -408,62 +424,123 @@ func validateFinetuneConfig(c *config.FineTuneConfig) error {
...
-		c.Config.Unsloth.WeightDecay = 0.01
+	if unsloth.Seed < 0 {
+		return errors.New("config.unsloth.seed must be zero or greater")
 	}
-	if c.Config.Unsloth.LrSchedulerType == "" {
-		c.Config.Unsloth.LrSchedulerType = "linear"
+
+	if !isSupportedUnslothQuantization(c.Output.Quantize) {
+		return errors.Errorf("output quantization %s is not supported", c.Output.Quantize)
+	}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go]: Fixed in `6f4e922`. Empty and whitespace-only `output.quantize` values now return `output.quantize is not defined`, while unsupported values are quoted in the error. Added tests for both blank cases and the quoted unsupported message.

File: pkg/aikit2llb/finetune/convert.go [hunk 4]
```diff
@@ -2,81 +2,216 @@ package finetune
...
 
-	return scratch
+func nvidiaDriverInstallOptions(version string) []llb.RunOption {
+	resolveVersion := `VERSION=$(sed -n 's/.*NVIDIA UNIX x86_64 Kernel Module  \([0-9]\+\.[0-9]\+\.[0-9]\+\).*/\1/p' /proc/driver/nvidia/version)`
+	if version != "" {
+		resolveVersion = "VERSION=" + shellQuote(version)
+	}
+
+	command := fmt.Sprintf(`%s &&
+test -n "$VERSION" &&
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/aikit2llb/finetune/convert.go]: Fixed in `6f4e922`. Automatic driver-version discovery now funnels missing or unparsable procfs data into an actionable error that explains both remedies: expose the host NVIDIA procfs entry or pass `--build-arg nvidiaDriverVersion=<major.minor.patch>`. Added generated-command coverage and removed the silent `test -n` check.

**Comment by sozercan** [comment]
@codex review

File: pkg/build/build.go [hunk 5]
```diff
@@ -384,9 +390,19 @@ func getBuildArg(opts map[string]string, k string) string {
 	return ""
 }
 
+func parseFineTuneBuildOptions(opts map[string]string) (finetune.Options, error) {
+	driverVersion := strings.TrimSpace(getBuildArg(opts, "nvidiaDriverVersion"))
+	if driverVersion != "" && !nvidiaDriverVersionPattern.MatchString(driverVersion) {
+		return finetune.Options{}, errors.Errorf("nvidiaDriverVersion %q must use major.minor.patch format", driverVersion)
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go]: Fixed. `nvidiaDriverVersion` now accepts both `major.minor` and `major.minor.patch`, so WSL values such as `572.83` are valid. The parser tests cover the two-component form.

File: pkg/build/build.go [hunk 6]
```diff
@@ -36,6 +38,8 @@ const (
 	keyCacheImports   = "cache-imports"
 )
 
+var nvidiaDriverVersionPattern = regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+$`)
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go]: Fixed. The driver-version validation accepts both two- and three-component numeric versions, including the WSL example `572.83`, with regression coverage in `Test_parseFineTuneBuildOptions`.

File: pkg/build/build.go [hunk 7]
```diff
@@ -408,62 +428,126 @@ func validateFinetuneConfig(c *config.FineTuneConfig) error {
...
+		return true
+	default:
+		return false
 	}
-	return c
+}
+
+// isSupportedUnslothQuantization mirrors the quantization methods supported by the pinned Unsloth integration.
+func isSupportedUnslothQuantization(quantization string) bool {
+	switch strings.ToLower(quantization) {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go:530]: Fixed. Quantization identifiers are canonicalized to lowercase during validation before they are serialized for export, so accepted uppercase input cannot reach Unsloth unchanged.

File: pkg/build/build.go [hunk 8]
```diff
@@ -408,62 +428,126 @@ func validateFinetuneConfig(c *config.FineTuneConfig) error {
...
+func isSupportedUnslothOptimizer(optimizer string) bool {
+	switch optimizer {
+	case "adamw_torch", "adamw_torch_fused", "adamw_torch_xla", "adamw_torch_npu_fused", "adamw_apex_fused",
+		"adafactor", "adamw_anyprecision", "adamw_torch_4bit", "adamw_torch_8bit", "ademamix", "sgd", "adagrad",
+		"adamw_bnb_8bit", "adamw_8bit", "ademamix_8bit", "lion_8bit", "lion_32bit", "paged_adamw_32bit",
+		"paged_adamw_8bit", "paged_ademamix_32bit", "paged_ademamix_8bit", "paged_lion_32bit", "paged_lion_8bit",
+		"rmsprop", "rmsprop_bnb", "rmsprop_bnb_8bit", "rmsprop_bnb_32bit", "galore_adamw", "galore_adamw_8bit",
+		"galore_adafactor", "galore_adamw_layerwise", "galore_adamw_8bit_layerwise", "galore_adafactor_layerwise",
+		"lomo", "adalomo", "grokadamw", "schedule_free_radam", "schedule_free_adamw", "schedule_free_sgd",
+		"apollo_adamw", "apollo_adamw_layerwise", "stable_adamw":
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go]: Fixed. The supported optimizer set is restricted to implementations available in the frozen environment, so configurations requiring absent optional packages are rejected before GPU work.

File: .github/workflows/test-finetune.yaml [hunk 9]
```diff
@@ -2,92 +2,429 @@ name: docker-test-finetune
...
 
-      - run: docker system prune -a -f --volumes || true
+      - name: Run test model
+        shell: bash
+        run: |
+          set -euo pipefail
+          docker run \
+            --name "$CONTAINER_NAME" \
+            --detach \
+            --device "$NVIDIA_CDI_DEVICE" \
```
[the remark this answers is not in the document]
Comment (sozercan) [reply .github/workflows/test-finetune.yaml]: Fixed. BuildKit and Docker runtime CDI selectors are now separate. The workflow may use an on-demand builder kind, but inference always uses the concrete runtime selector `nvidia.com/gpu=0`.

[the remark this answers is not in the document]
Comment (sozercan) [reply .github/workflows/test-finetune.yaml]: Fixed. The workflow keeps the BuildKit CDI selector independent from the Docker runtime selector, so a bare on-demand kind is never passed to `docker run`.

File: pkg/finetune/target_unsloth.py [hunk 10]
```diff
@@ -56,67 +27,284 @@
...
+    if dependencies is None:
+        dependencies = load_export_dependencies()
+
+    cfg = unsloth_config(export_config)
+    trained_model_path = Path(trained_model_directory)
+    export_path = Path(export_directory)
+
+    model, tokenizer = dependencies.fast_language_model.from_pretrained(
+        model_name=str(trained_model_path),
+        max_seq_length=cfg["maxSeqLength"],
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:381]: Fixed. Training resolves the non-quantized export repository to an immutable Hugging Face SHA and pins it in the PEFT config. Export downloads that exact snapshot, rewrites the adapter to the local immutable path, and reloads with `local_files_only=True`. Both final-head full and smoke GPU runs pass.

File: pkg/build/build.go [hunk 11]
```diff
@@ -408,62 +435,126 @@ func validateFinetuneConfig(c *config.FineTuneConfig) error {
...
+	default:
+		return false
+	}
+}
+
+// isSupportedUnslothScheduler mirrors SchedulerType from the pinned Transformers version.
+func isSupportedUnslothScheduler(scheduler string) bool {
+	switch scheduler {
+	case "linear", "cosine", "cosine_with_restarts", "polynomial", "constant", "constant_with_warmup",
+		"inverse_sqrt", "reduce_lr_on_plateau", "cosine_with_min_lr", "cosine_warmup_with_min_lr",
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/build/build.go]: Fixed. Scheduler validation is restricted to the API-compatible subset that works without unavailable scheduler kwargs or evaluation settings, so unsupported choices fail before model loading.

File: .github/workflows/test-finetune.yaml [hunk 12]
```diff
@@ -2,92 +2,447 @@ name: docker-test-finetune
...
+          if ! grep -Fq 'org.mobyproject.buildkit.device.autoallow:' "$cdi_spec"; then
+            if grep -q '^annotations:' "$cdi_spec"; then
+              sudo sed -i '/^annotations:/a\  org.mobyproject.buildkit.device.autoallow: "true"' "$cdi_spec"
+            else
+              sudo sed -i '/^kind:/a\annotations:\n  org.mobyproject.buildkit.device.autoallow: "true"' "$cdi_spec"
+            fi
+          fi
+          sudo install -d -m 0755 /etc/cdi
+          sudo install -m 0644 "$cdi_spec" /etc/cdi/nvidia.yaml
+          sudo rm -f "$cdi_spec" /var/run/cdi/nvidia.yaml
```
[the remark this answers is not in the document]
Comment (sozercan) [reply .github/workflows/test-finetune.yaml]: Fixed. The workflow no longer rewrites or removes host-global CDI files. Runner provisioning owns CDI refresh; the job only verifies the runtime and builder device inventory.

File: pkg/aikit2llb/finetune/convert.go [hunk 13]
```diff
@@ -8,86 +8,164 @@ import (
...
+	return llb.Scratch().File(llb.Copy(state, inputFile, outputFile, copyOpts...)), nil
+}
+
+func gpuCacheKey(opts Options) string {
+	if opts.NVIDIADriverVersion != "" {
+		return "driver:" + opts.NVIDIADriverVersion
+	}
+	if opts.BuildSessionID != "" {
+		return "session:" + opts.BuildSessionID
+	}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/aikit2llb/finetune/convert.go:147]: Fixed. Cross-session GPU result caching now requires a canonical immutable `GPU-...` or `MIG-...` CDI UUID paired with the driver version. Mutable selectors (`0`, `all`, type-index, or on-demand) fall back to the BuildKit session key. The workflow prefers a canonical builder-advertised UUID, rejects malformed aliases in every fallback pass, and the cache behavior has regression tests. Final-head full and smoke GPU runs pass.

**Comment by sozercan** [comment]
@codex review

**Comment by sozercan** [comment]
@codex review
